# Лабораторная работа №6

## Цель работы

Освоить принципы использования механизма обработки исключительных ситуаций при считывании/записи в файл на примере функции для сохранения лога операций и чтения настроек для работы калькулятора из файла. 

## Комментарии по выполнению

Работу можно структурировать на следующие части:

- Проанализировать ситуации работы с файлом, в которых может возникнуть исключение и реализовать обработку этих исключительных ситуаций с помощью базового инструментария, показанного в конспекте курса или по ссылкам (официальная документация, русско-язычный ресурс и ещё один по обработке исключений). 

Опишем конкретные аспекты задания ниже:

### - Анализ мест в коде с исключительными ситуациями

Можно вместо образца борда по ссылке в начале задания отлавливать исключения (errors) в функционале вашего калькулятора из предыдущих лабораторных работ, особенно в той, где надо было читать все необходимые и выраженния для вычислений из файла.

Исключительная ситуация может возникнуть на этапе работы с файлом (чтение, запись). Программа может не считать файл с настройками например, из-за ограниченного набора прав пользователя, запустившего данную программу или каких-либо настроек других ОС. Эти ситуации мы можем обработать с помощью блока (см. рабочий пример в стартовом борде):
```
try
    pass # какое-то выражение, возможно, поднимающее исключение  
except Exception:
   print('Исключение возникло') # обработка исключения 
else:
   # блок, выполняющийся, если исключения не было
finally:
   # блок, который выполняется в всегда независимо от того, было ли исключение или нет
```

```
def __enter__(self):
    # Открываем файл на чтение
    try:
        self.file = open(self.file_path, 'r', encoding='utf-8')
    except OSError as e:
        print(f"Ошибка открытия файла: {e}")
    else:
        return self.file
```

```
def cmd_read():
    """
    Главная функция для запуска калькулятора.
    
    Запрашивает у пользователя ввод чисел и операции.
    
    Выводит результат вычисления.
    """
    try:
        numbers = list(map(str, input("Введите числа ")).split(" "))
        operand = str(input("Введите арифметическая операцию "))
        tolerance = str(input("Введите точность "))
        print("Результат выражения ", calculate(numbers, operand, tolerance))
    except Exception as e:
        print(f"Ошибка: {e}")
    
def file_read():
    """
    Главная функция для запуска калькулятора.
    
    Читает из файла все выражения.
    
    Выводит результат вычисления.
    """
    file_path = "prog\expressions.txt"  # Имя файла с арифметическими выражениями
    work_numbers = "0123456789"
    work_opertand = "+ - * /"
    
    # Обрабатываем файл с помощью менеджера контекста и генератора
    try:
        with BatchCalculatorContextManager(file_path) as file:
            for expression in expression_generator(file):
                number = ""
                numbers = []
                for char in expression:
                    if char in work_numbers:
                        number += char
                    elif char == " ":
                        next
                    elif char in work_opertand:
                        operand = char
                        numbers.append(int(number))
                        number = ""
                    else:
                        print("Неправильно записано выражение в файле")
                numbers.append(int(number))
                result = calculate(numbers, operand, "")
                print(f"{expression} = {result}")
    except Exception as e:
        print(f"Ошибка при работе с файлом: {e}")
```
[ссылка на файл](https://github.com/ZabivakaXD/Herzen_curse_2/blob/main/prog/Calculate.py)